File: toolchain/rust/src/repl.rs
//! Read Eval Print Loop.

use std::io;

use rustyline::{Editor, Result};

use crate::lexer::Lexer;
use crate::token::TokenKind;

const PROMPT: &str = ">> ";

/// Starts the REPL.
pub fn start(mut writer: impl io::Write) -> Result<()> {
    let mut line_reader = Editor::<()>::new()?;
    while let Ok(line) = line_reader.readline(PROMPT) {
        let mut lexer = Lexer::from_text(&line);
        while let Some(tok) = lexer.next_token() {
            if tok.kind == TokenKind::Eof {
                break;
            }
            write!(&mut writer, "{:?}", tok).expect("failed to write");
            writer.flush()?;
        }
    }
    Ok(())
}

File: toolchain/go/repl/repl.go
// Package repl implement the REPL.
package repl

import (
	"bufio"
	"fmt"
	"github/com/styvane/monkey/lexer"
	"github/com/styvane/monkey/token"
	"io"
)

const PROMPT = ">>> "

func Start(in io.Reader, out io.Writer) {
	scanner := bufio.NewScanner(in)
	for {
		fmt.Fprintf(out, PROMPT)
		scanned := scanner.Scan()
		if !scanned {
			return
		}
		line := scanner.Text()
		l := lexer.New(line)
		for tok := l.NextToken(); tok.Kind != token.EOF; tok = l.NextToken() {
			fmt.Fprintf(out, "%+v\n", tok)
		}

	}
}
